fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -252,34 +267,45 @@ private static void setDeprecated(DescriptorProtos.FieldDescriptorProto.Builder
      */
     @Nonnull
     public static Descriptors.FileDescriptor addDefaultUnionIfMissing(Descriptors.FileDescriptor fileDescriptor) {
-        DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptor.toProto().toBuilder();
-        if (MetaDataProtoEditor.hasUnion(fileBuilder)) {
+        if (MetaDataProtoEditor.hasUnion(fileDescriptor)) {
             return fileDescriptor;
         }
-        MetaDataProtoEditor.addDefaultUnion(fileBuilder);
+        DescriptorProtos.FileDescriptorProto fileDescriptorProto = fileDescriptor.toProto();
+        DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptorProto.toBuilder();
+        fileBuilder.addMessageType(MetaDataProtoEditor.createDefaultUnion(fileDescriptorProto));
         try {
             return Descriptors.FileDescriptor.buildFrom(fileBuilder.build(), fileDescriptor.getDependencies().toArray(new Descriptors.FileDescriptor[0]));
         } catch (Descriptors.DescriptorValidationException e) {
             throw new MetaDataException("Failed to add a default union", e);
         }
     }
 
+    /**
+     * Creates a default union descriptor for the given file descriptor. It adds all of the non-{@code NESTED} message types to the union.
+     * @param fileDescriptorProto the file descriptor to create a union for
+     * @return the builder for the union
+     */
     @Nonnull
-    private static void addDefaultUnion(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileDescriptorBuilder) {
+    public static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull DescriptorProtos.FileDescriptorProto fileDescriptorProto) {
         DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
         unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
-        for (DescriptorProtos.DescriptorProto messageType : fileDescriptorBuilder.getMessageTypeList()) {
+        for (DescriptorProtos.DescriptorProto messageType : fileDescriptorProto.getMessageTypeList()) {
             RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType);
             if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
-                addFieldToUnion(unionMessageType, messageType);
+                addFieldToUnion(unionMessageType, messageType.getName(), fileDescriptorProto.getPackage() + "." + messageType.getName());
```

Comment:
How about we enforce that the package is never empty, by rejecting such a proto?

Reply:
Maybe? Are empty package names legal in protobuf?

Reply:
In any case, I think passing the file descriptor instead of the file descriptor proto and calling `getFullName` is a better solution.

Reply:
> In any case, I think passing the file descriptor instead of the file descriptor proto and calling getFullName is a better solution.

I don't think `getFullName` works. It will return the full name of the file, right? We need the package here.

>Maybe? Are empty package names legal in protobuf?
 
I've never seen it missing, but it's an optional field.

Reply:
> parentFileDescriptor.getFullName()
> result = test_no_union.proto

> parentFileDescriptor.getPackage()
> result = com.apple.foundationdb.record.testnounion

Reply:
I meant `getFullName` on the message descriptors. Those should give back the fully-qualified paths.

Reply:
I see, yeah that's right. Anyways, IMO, the two are not really that different. Changed it to `FileDescriptor` and used `getFullName`.